### PR TITLE
Make a compat story for confidential token ids

### PR DESCRIPTION
**Pull request by cbeck88**
In this story, the transaction builder can build old-style and
new-style Txs.

### CODE REVIEW

**Review by eranrund — COMMENTED** [review]
I went over this and overall I think it looks good.
I am wondering if block versions should not be represented in Rust as an Enum, with helper methods such as `uses_opening_proof` or something like that. In the long term I think this will result in more readable code: instead of having if block_version > 2 && block_version < 4 (for example), we will have `block_version.uses_opening_proof()`.
This does get annoying due to Prost, but its similar to TokenId in that regard.

Another thing I want to point out is the amount of boolean arguments to `sign_with_balance_check`. It might be worth moving to passing a struct to it with configuration options, or doing a builder pattern. This can happen later.

I don't remember what issues we previously ran into when bumping the block version, it might've broken ledger syncing or something like that but its not necessarily a concern for this particular PR.

Following what we talked about in Slack, I remain of the belief that 90% of the changes here  have nothing to do with block_version=2, and as such this can be constructed as a standalone PR that goes into `master`. There are a ton of test changes that are very hard to follow, and it would be nice to know for sure that nothing in them has changed to accommodate the new block version.
This will also reduce the diff size in the `confidential-token-ids` PR and I will continue to push on doing whatever we can to make that change as incremental as possible.

**Comment by cbeck88** [comment]
The thing is that, this PR is fixing what are considered flaws in confidential-token-ids that would cause breaking changes, and it's also changing aspects of how the signature validation works. So I think it might require more reviewer effort and not less if we don't merge this.

We could try to split parts of this out though, like making the transaction builder take a block version argument

**Comment by eranrund** [comment]
> We could try to split parts of this out though, like making the transaction builder take a block version argument

Yes, that is what I'm talking about. That + all the new loop stuff in the tests that doesn't have behavior that is different based on block version.

**Comment by cbeck88** [comment]
Yeah but it does try a couple token ids, and until the token ids PR, the token ids don't exist?

Maybe we could still port it to master and just ignore the token ids for now I guess?

**Comment by eranrund** [comment]
(discussion continued over Slack)
